### PR TITLE
[SYCLomatic] Custom the helper function usage in the migrated code by rule file

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

**Review by AndyCHHuang — APPROVED** [review]
LGTM

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 1]
```diff
@@ -1225,15 +1226,30 @@ std::string DpctGlobalInfo::getDefaultQueue(const Stmt *S) {
...
+      MapNames::CustomHelperFunctionMap.find(HelperFuncCatalog::DefaultQueue);
+  if (Iter != MapNames::CustomHelperFunctionMap.end()) {
+    return Iter->second;
+  }
+  if (DpctGlobalInfo::useSYCLCompat())
+    return DefaultQueueFreeFunc;
+  if (DpctGlobalInfo::getUsmLevel() == UsmLevel::UL_None)
+    return OutOfOrderQueueFreeFunc;
+  return InOrderQueueFreeFunc;
+}
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]: Use a lambda like left side.

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 2]
```diff
@@ -1225,15 +1226,30 @@ std::string DpctGlobalInfo::getDefaultQueue(const Stmt *S) {
...
+  auto Iter =
+      MapNames::CustomHelperFunctionMap.find(HelperFuncCatalog::DefaultQueue);
+  if (Iter != MapNames::CustomHelperFunctionMap.end()) {
+    return Iter->second;
+  }
+  if (DpctGlobalInfo::useSYCLCompat())
+    return DefaultQueueFreeFunc;
+  if (DpctGlobalInfo::getUsmLevel() == UsmLevel::UL_None)
+    return OutOfOrderQueueFreeFunc;
+  return InOrderQueueFreeFunc;
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]: Can call getDefaultQueueMemFuncName

File: clang/test/dpct/user_defined_rule2.cu [hunk 3]
```diff
@@ -0,0 +1,37 @@
+// RUN: dpct --out-root %T/user_defined_rule2 %s --cuda-include-path="%cuda-path/include" --rule-file %S/ipex_xpu.yaml --format-range=none
+// RUN: FileCheck --input-file %T/user_defined_rule2/user_defined_rule2.dp.cpp --match-full-lines %s
+// RUN: %if build_lit %{icpx -c -fsycl -DBUILD_TEST  %T/user_defined_rule2/user_defined_rule2.dp.cpp -o %T/user_defined_rule2/user_defined_rule2.dp.o %}
+
+#ifndef BUILD_TEST
+
+__global__ void foo1_kernel() {}
+void foo1() {
+  // CHECK: dpct::static_cast<sycl::queue&>(c10::xpu::getCurrentXPUStream()).parallel_for(
```
Comment (zhiweij1) [inline clang/test/dpct/user_defined_rule2.cu]:
```suggestion
  // CHECK: static_cast<sycl::queue&>(c10::xpu::getCurrentXPUStream()).parallel_for(
```

File: clang/test/dpct/user_defined_rule2.cu [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  foo1_kernel<<<1, 1>>>();
+}
+
+__global__ void foo2_kernel(double *d) {}
+
+void foo2() {
+  double *d;
+  // CHECK: d = sycl::malloc_device<double>(1, dpct::static_cast<sycl::queue&>(c10::xpu::getCurrentXPUStream()));
+  // CHECK-NEXT: {
+  // CHECK-NEXT:   dpct::has_capability_or_fail(static_cast<sycl::queue&>(c10::xpu::getCurrentXPUStream()).get_device(), {sycl::aspect::fp64});
```
Comment (zhiweij1) [inline clang/test/dpct/user_defined_rule2.cu]: Add this free function into SYCLCompat 1st, then change the migration from member call to free func call.
Then we can remove rule2, and if we have rule1, here we can use rule1.Out.get_device()

File: clang/test/dpct/ipex_xpu.yaml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+---
+- Rule: rule1
+  Kind: HelperFunction
+  Priority: Takeover
+  In: DefaultQueue
+  Out: static_cast<sycl::queue&>(c10::xpu::getCurrentXPUStream())
+  Includes: [""]
```
Comment (zhiweij1) [inline clang/test/dpct/ipex_xpu.yaml]:
```suggestion
```

**Comment by zhiweij1** [comment]
SYCLCompat API added https://github.com/intel/llvm/pull/15717

File: clang/test/dpct/xpu.yaml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+---
+- Rule: rule1
+  Kind: HelperFunction
+  Priority: Takeover
+  In: DefaultQueue
```
Comment (zhimingwang36) [inline clang/test/dpct/xpu.yaml]: get_in_order_queue()/get_out_of_order_queue

File: clang/lib/DPCT/Rules.cpp [hunk 7]
```diff
@@ -277,6 +277,25 @@ void registerPatternRewriterRule(MetaRuleObject &R) {
       R.BuildScriptSyntax, R.Priority));
 }
 
+void registerHelperFunctionRule(MetaRuleObject &R) {
+  if ((R.In == "get_default_queue" || R.In == "get_in_order_queue" ||
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Rules.cpp]: If another helper function is specified, please emit the warning "The rule named xxx (kind: helperfunction) is ignored, for the API specified in 'In' field is not supported to be customized."

File: clang/lib/DPCT/MapNames.cpp [hunk 8]
```diff
@@ -4597,6 +4597,8 @@ MapNames::MapTy TextureRule::ResourceTypeNames{{"devPtr", "data_ptr"},
                                                {"numChannels", "channel_num"}};
 
 std::vector<MetaRuleObject::PatternRewriter> MapNames::PatternRewriters;
+std::map<clang::dpct::HelperFuncCatalog, std::string>
```
Comment (zhimingwang36) [inline clang/lib/DPCT/MapNames.cpp:4603]: Please add comment for the new added map.

File: clang/lib/DPCT/AnalysisInfo.h [hunk 9]
```diff
@@ -1643,6 +1646,7 @@ class DpctGlobalInfo {
   static std::vector<std::pair<std::string, std::vector<std::string>>>
       CodePinDumpFuncDepsVec;
   static std::unordered_set<std::string> NeedParenAPISet;
+  static std::unordered_set<std::string> CustomHelperFunctionAddtionalIncludes;
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h:1650]: Cleanup the content before 2nd migration

**Review by zhimingwang36 — APPROVED** [review]

**Pull request merged**